vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java

```diff
@@ -26,17 +28,53 @@ public class MixedTensor implements Tensor {
 
     /** The dimension specification for this tensor */
     private final TensorType type;
+    private final int blockSize; // aka dense subspace size
 
-    /** The list of cells in the tensor */
-    private final List<Cell> cells;
+    static final class DenseBlock {
+        final TensorAddress sparseAddr;
+        final double[] cells;
+        DenseBlock(TensorAddress sparseAddr, double[] cells) {
+            this.sparseAddr = sparseAddr;
+            this.cells = cells;
+        }
+        @Override public int hashCode() {
+            return Objects.hash(sparseAddr, cells);
+        }
+        @Override public boolean equals(Object other) {
+            if (other instanceof DenseBlock o) {
+                return sparseAddr.equals(o.sparseAddr) && Arrays.equals(cells, o.cells);
+            }
+            return false;
+        }
+    }
+
+    /** The cells in the tensor */
+    private final List<DenseBlock> cellBlocks;
 
     /** An index structure over the cell list */
     private final Index index;
 
-    private MixedTensor(TensorType type, List<Cell> cells, Index index) {
+    private MixedTensor(TensorType type, List<DenseBlock> cellBlocks, Index index) {
         this.type = type;
-        this.cells = List.copyOf(cells);
+        this.blockSize = index.denseSubspaceSize();
+        this.cellBlocks = List.copyOf(cellBlocks);
         this.index = index;
+        if (this.blockSize < 1) {
+            throw new IllegalStateException("invalid dense subspace size " + blockSize);
+        }
+        long count = 0;
+        for (var block : this.cellBlocks) {
+            if (index.sparseMap.get(block.sparseAddr) != count) {
+                throw new IllegalStateException("map vs list mismatch");
```

Comment:
Consider outputting the sparse address so its easier to pinpoint the issue if this happens.